# What's in the database?


In [1]:
%reload_ext sql
%sql postgresql+psycopg2://adrik:root1234@localhost:5432/adrik

In [2]:
%%sql

SELECT count(*) 
  FROM stackoverflow;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(45238,)]

In [5]:
%%sql

SELECT *
FROM fortune500
LIMIT 5;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
5 rows affected.


[(1, 'Walmart', 'Wal-Mart Stores, Inc.', 'WMT  ', 'http://www.walmart.com', 'Bentonville, AR', 'Retailing', 'General Merchandisers', 2300000, 485873, 0.8, Decimal('13643'), -7.2, Decimal('198825'), Decimal('77798')),
 (2, 'Berkshire Hathaway', 'Berkshire Hathaway Inc.', 'BRKA ', 'http://www.berkshirehathaway.com', 'Omaha, NE', 'Financials', 'Insurance: Property and Casualty (Stock)', 367700, 223604, 6.1, Decimal('24074'), 0.0, Decimal('620854'), Decimal('283001')),
 (3, 'Apple', 'Apple, Inc.', 'AAPL ', 'http://www.apple.com', 'Cupertino, CA', 'Technology', 'Computers, Office Equipment', 116000, 215639, -7.7, Decimal('45687'), -14.4, Decimal('321686'), Decimal('128249')),
 (4, 'Exxon Mobil', 'Exxon Mobil Corporation', 'XOM  ', 'http://www.exxonmobil.com', 'Irving, TX', 'Energy', 'Petroleum Refining', 72700, 205004, -16.7, Decimal('7840'), -51.5, Decimal('330314'), Decimal('167325')),
 (5, 'McKesson', 'McKesson Corporation', 'MCK  ', 'http://www.mckesson.com', 'San Francisco, CA', 'Wholesalers', 'Wholesalers: Health Care', 68000, 192487, 6.2, Decimal('2258'), 53.0, Decimal('56563'), Decimal('8924'))]

> stackoverflow has the most rows; fortune500 has the most columns

### Count missing values


In [6]:
%%sql

-- Select the count of ticker, 
-- subtract from the total number of rows, 
-- and alias as missing
SELECT count(*) - COUNT(ticker) AS missing
  FROM fortune500;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(32,)]

In [7]:
%%sql

-- Select the count of profits_change, 
-- subtract from total number of rows, and alias as missing
SELECT COUNT(*) - COUNT(profits_change) AS missing
FROM fortune500

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(63,)]

In [8]:
%%sql

-- Select the count of industry, 
-- subtract from total number of rows, and alias as missing
SELECT COUNT(*) - COUNT(industry) AS missing
FROM fortune500

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(13,)]

### Join tables


In [9]:
%%sql

SELECT company.name
-- Table(s) to select from
  FROM company
       INNER JOIN fortune500
       ON company.ticker=fortune500.ticker;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
8 rows affected.


[('Apple Incorporated',),
 ('Amazon.com Inc',),
 ('Alphabet',),
 ('Microsoft Corp.',),
 ('International Business Machines Corporation',),
 ('PayPal Holdings Incorporated',),
 ('eBay, Inc.',),
 ('Adobe Systems Incorporated',)]

## The keys to the database
### Foreign keys
> Why can't the tag column in the tag_type table be a foreign key that references the tag column in the stackoverflow table?

> stackoverflow.tag contains duplicate values

### Read an entity relationship diagram

#### What is the most common stackoverflow tag_type?


In [10]:
%%sql

-- Count the number of tags with each type
SELECT type, COUNT(tag) AS count
  FROM tag_type
 -- To get the count for each type, what do you need to do?
 GROUP BY type
 -- Order the results with the most common
 -- tag types listed first
 ORDER BY count DESC;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[('cloud', 31),
 ('database', 6),
 ('payment', 5),
 ('mobile-os', 4),
 ('api', 4),
 ('company', 4),
 ('storage', 2),
 ('os', 2),
 ('spreadsheet', 2),
 ('identity', 1)]

In [11]:
%%sql

SELECT company.name, tag_type.tag, tag_type.type
  FROM company
  	   -- Join to the tag_company table
       INNER JOIN tag_company 
       ON company.id = tag_company.company_id
       -- Join to the tag_type table
       INNER JOIN tag_type
       ON tag_company.tag = tag_type.tag
  -- Filter to most common type
  WHERE type='cloud';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
31 rows affected.


[('Amazon Web Services', 'amazon-cloudformation', 'cloud'),
 ('Amazon Web Services', 'amazon-cloudfront', 'cloud'),
 ('Amazon Web Services', 'amazon-cloudsearch', 'cloud'),
 ('Amazon Web Services', 'amazon-cloudwatch', 'cloud'),
 ('Amazon Web Services', 'amazon-cognito', 'cloud'),
 ('Amazon Web Services', 'amazon-data-pipeline', 'cloud'),
 ('Amazon Web Services', 'amazon-dynamodb', 'cloud'),
 ('Amazon Web Services', 'amazon-ebs', 'cloud'),
 ('Amazon Web Services', 'amazon-ec2', 'cloud'),
 ('Amazon Web Services', 'amazon-ecs', 'cloud'),
 ('Amazon Web Services', 'amazon-elastic-beanstalk', 'cloud'),
 ('Amazon Web Services', 'amazon-elasticache', 'cloud'),
 ('Amazon Web Services', 'amazon-elb', 'cloud'),
 ('Amazon Web Services', 'amazon-emr', 'cloud'),
 ('Amazon Web Services', 'amazon-glacier', 'cloud'),
 ('Amazon Web Services', 'amazon-kinesis', 'cloud'),
 ('Amazon Web Services', 'amazon-lambda', 'cloud'),
 ('Amazon Web Services', 'amazon-rds-aurora', 'cloud'),
 ('Amazon Web Services', 'amazon-rds', 'cloud'),
 ('Amazon Web Services', 'amazon-redshift', 'cloud'),
 ('Amazon Web Services', 'amazon-route53', 'cloud'),
 ('Amazon Web Services', 'amazon-s3', 'cloud'),
 ('Amazon Web Services', 'amazon-ses', 'cloud'),
 ('Amazon Web Services', 'amazon-simpledb', 'cloud'),
 ('Amazon Web Services', 'amazon-sns', 'cloud'),
 ('Amazon Web Services', 'amazon-sqs', 'cloud'),
 ('Amazon Web Services', 'amazon-swf', 'cloud'),
 ('Amazon Web Services', 'amazon-vpc', 'cloud'),
 ('Amazon Web Services', 'amazon-web-services', 'cloud'),
 ('Microsoft Corp.', 'azure', 'cloud'),
 ('Dropbox', 'dropbox', 'cloud')]

### Coalesce

In [14]:
%%sql

SELECT COALESCE(industry, sector, 'Unknown') AS industry2,
       -- Don't forget to count!
       COUNT(*) 
  FROM fortune500 
-- Group by what? (What are you counting by?)
 GROUP BY industry2
-- Order results to see most common first
 ORDER BY count DESC
-- Limit results to get just the one value you want
 LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('Utilities: Gas and Electric', 22),
 ('Specialty Retailers: Other', 21),
 ('Commercial Banks', 19),
 ('Insurance: Property and Casualty (Stock)', 19),
 ('Chemicals', 14),
 ('Food Consumer Products', 13),
 ('Aerospace and Defense', 12),
 ('Diversified Financials', 12),
 ('Telecommunications', 10),
 ('Insurance: Life, Health (Stock)', 10),
 ('Energy', 10),
 ('Petroleum Refining', 10),
 ('Pharmaceuticals', 10),
 ('Financial Data Services', 9),
 ('Semiconductors and Other Electronic Components', 9),
 ('General Merchandisers', 9),
 ('Wholesalers: Diversified', 9),
 ('Motor Vehicles and Parts', 9),
 ('Internet Services and Retailing', 9),
 ('Automotive Retailing, Services', 9)]

### Coalesce with a self-join


In [15]:
%%sql

SELECT company_original.name, fortune500.title, fortune500.rank 
  -- Start with original company information
  FROM company AS company_original
       -- Join to another copy of company with parent
       -- company information
	   LEFT JOIN company AS company_parent
       ON company_original.parent_id = company_parent.id 
       -- Join to fortune500, only keep rows that match
       INNER JOIN fortune500 
       -- Use parent ticker if there is one, 
       -- otherwise original ticker
       ON coalesce(company_original.ticker, 
                   company_parent.ticker) = 
             fortune500.ticker
 -- For clarity, order by rank
 ORDER BY rank; 

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[('Apple Incorporated', 'Apple', 3),
 ('Amazon.com Inc', 'Amazon.com', 12),
 ('Amazon Web Services', 'Amazon.com', 12),
 ('Alphabet', 'Alphabet', 27),
 ('Google LLC', 'Alphabet', 27),
 ('Microsoft Corp.', 'Microsoft', 28),
 ('International Business Machines Corporation', 'IBM', 32),
 ('PayPal Holdings Incorporated', 'PayPal Holdings', 264),
 ('eBay, Inc.', 'eBay', 310),
 ('Adobe Systems Incorporated', 'Adobe Systems', 443)]

## Column types and constraints
### Effects of casting



In [21]:
%%sql

-- Select the original value
SELECT profits_change, 
	   -- Cast profits_change
       CAST(profits_change AS integer) AS profits_change_int
  FROM fortune500
  LIMIT 10;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[(-7.2, -7),
 (0.0, 0),
 (-14.4, -14),
 (-51.5, -52),
 (53.0, 53),
 (20.7, 21),
 (1.5, 2),
 (-2.7, -3),
 (-2.8, -3),
 (-37.7, -38)]

In [20]:
%%sql 

-- Select the original value
SELECT profits_change, 
	   -- Cast profits_change
       profits_change::integer AS profits_change_int
  FROM fortune500
  LIMIT 10;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[(-7.2, -7),
 (0.0, 0),
 (-14.4, -14),
 (-51.5, -52),
 (53.0, 53),
 (20.7, 21),
 (1.5, 2),
 (-2.7, -3),
 (-2.8, -3),
 (-37.7, -38)]

In [18]:
%%sql

-- Divide 10 by 3
SELECT 10/3, 
       -- Cast 10 as numeric and divide by 3
       10::numeric/3;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(3, Decimal('3.3333333333333333'))]

In [19]:
%%sql

SELECT '3.2'::numeric,
       '-123'::numeric,
       '1e3'::numeric,
       '1e-3'::numeric,
       '02314'::numeric,
       '0002'::numeric;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(Decimal('3.2'), Decimal('-123'), Decimal('1000'), Decimal('0.001'), Decimal('2314'), Decimal('2'))]

### Summarize the distribution of numeric values

#### Was 2017 a good or bad year for revenue of Fortune 500 companies? Examine how revenue changed from 2016 to 2017 by first looking at the distribution of revenues_change and then counting companies whose revenue increased.


In [26]:
%%sql

-- Select the count of each value of revenues_change
SELECT revenues_change, COUNT(revenues_change)
  FROM fortune500
 GROUP BY revenues_change
 -- order by the values of revenues_change
 ORDER BY revenues_change DESC
 LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[(197.3, 1),
 (122.1, 1),
 (115.9, 1),
 (94.5, 1),
 (92.6, 1),
 (91.8, 1),
 (83.0, 1),
 (78.6, 2),
 (73.0, 1),
 (59.5, 1),
 (54.2, 1),
 (48.1, 1),
 (45.6, 1),
 (44.4, 1),
 (41.7, 1),
 (40.1, 1),
 (38.5, 1),
 (37.9, 1),
 (37.2, 1),
 (33.7, 1)]

In [27]:
%%sql

-- Select the count of each revenues_change integer value
SELECT revenues_change::integer, COUNT(revenues_change)
  FROM fortune500
 GROUP BY revenues_change::integer
 -- order by the values of revenues_change
 ORDER BY revenues_change DESC
LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[(197, 1),
 (122, 1),
 (116, 1),
 (94, 1),
 (93, 1),
 (92, 1),
 (83, 1),
 (79, 2),
 (73, 1),
 (60, 1),
 (54, 1),
 (48, 1),
 (46, 1),
 (44, 1),
 (42, 1),
 (40, 1),
 (38, 2),
 (37, 1),
 (34, 1),
 (33, 1)]

In [28]:
%%sql
 
SELECT COUNT(*)
  FROM fortune500
 WHERE revenues_change >= 0;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(298,)]

> Examining distributions and counting observations of interest are two first steps in exploring data. 

## Numeric data types and summary functions
### Division

In [29]:
%%sql

-- Select average revenue per employee by sector
SELECT sector, 
       AVG(revenues/employees::numeric) AS avg_rev_employee
  FROM fortune500
 GROUP BY sector
 -- Use the column alias to order the results
 ORDER BY avg_rev_employee;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
21 rows affected.


[('Hotels, Restaurants & Leisure', Decimal('0.09498718151056814829')),
 ('Apparel', Decimal('0.27865942976680063809')),
 ('Food & Drug Stores', Decimal('0.30799950410060207070')),
 ('Motor Vehicles & Parts', Decimal('0.34252712424659522769')),
 ('Household Products', Decimal('0.35557338969595350494')),
 ('Retailing', Decimal('0.36019456092078082945')),
 ('Industrials', Decimal('0.36148543376146347043')),
 ('Aerospace & Defense', Decimal('0.36671499248628270960')),
 ('Transportation', Decimal('0.40365352477329589012')),
 ('Business Services', Decimal('0.42010994210166630178')),
 ('Technology', Decimal('0.57311847693359764307')),
 ('Chemicals', Decimal('0.59549976658074458280')),
 ('Telecommunications', Decimal('0.62958997279187493604')),
 ('Health Care', Decimal('0.79053286919681465178')),
 ('Media', Decimal('0.79561866565463184108')),
 ('Food, Beverages & Tobacco', Decimal('0.83088475939249850372')),
 ('Engineering & Construction', Decimal('0.86116376673874818555')),
 ('Wholesalers', Decimal('1.41324581189371202678')),
 ('Financials', Decimal('1.72638470140972568700')),
 ('Energy', Decimal('1.82601423325242362195')),
 ('Materials', Decimal('4.75758351554488369167'))]

### Explore with division


In [30]:
%%sql

-- Divide unanswered_count by question_count
SELECT unanswered_count/question_count::numeric AS computed_pct, 
       -- What are you comparing the above quantity to?
       unanswered_pct
  FROM stackoverflow
 -- Select rows where question_count is not 0
 WHERE question_count != 0
 LIMIT 10;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[(Decimal('0.46548476454293628809'), 0.001751857),
 (Decimal('0.38636363636363636364'), 0.000116972),
 (Decimal('0.39376770538243626062'), 5.8e-05),
 (Decimal('0.33189655172413793103'), 1.61e-05),
 (Decimal('0.42928571428571428571'), 0.000125312),
 (Decimal('0.34798961729250068974'), 0.012886449),
 (Decimal('0.35083862172255878033'), 0.007619406),
 (Decimal('0.30729166666666666667'), 1.23e-05),
 (Decimal('0.35428051001821493625'), 8.11e-05),
 (Decimal('0.38065776619993487463'), 0.000243743)]

> The values don't match. unanswered_pct is the percent of unanswered questions on Stack Overflow with the tag, not the percent of questions with the tag that are unanswered.

### Summarize numeric columns


In [31]:
%%sql

-- Select min, avg, max, and stddev of fortune500 profits
SELECT MIN(profits),
       AVG(profits),
       MAX(profits),
       STDDEV(profits)
  FROM fortune500;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(Decimal('-6177'), Decimal('1783.4753507014028056'), Decimal('45687'), Decimal('3940.495363490788'))]

In [32]:
%%sql

-- Select sector and summary measures of fortune500 profits
SELECT sector,
       MIN(profits),
       AVG(profits),
       MAX(profits),
       STDDEV(profits)
  FROM fortune500
 -- What to group by?
 GROUP BY sector
 -- Order by the average profits
 ORDER BY avg;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
21 rows affected.


[('Energy', Decimal('-6177'), Decimal('10.4446428571428571'), Decimal('7840'), Decimal('2264.572142925951')),
 ('Materials', Decimal('-440'), Decimal('272.4684210526315789'), Decimal('1027'), Decimal('406.632781447055')),
 ('Engineering & Construction', Decimal('15'), Decimal('390.1692307692307692'), Decimal('911.8'), Decimal('277.665120197620')),
 ('Wholesalers', Decimal('-199.4'), Decimal('391.2793103448275862'), Decimal('2258'), Decimal('532.171183776766')),
 ('Retailing', Decimal('-2221'), Decimal('991.7851063829787234'), Decimal('13643'), Decimal('2348.342559077222')),
 ('Chemicals', Decimal('-3.9'), Decimal('1137.0214285714285714'), Decimal('4318'), Decimal('1129.752304492226')),
 ('Business Services', Decimal('57.2'), Decimal('1155.3550000000000000'), Decimal('5991'), Decimal('1454.360686992199')),
 ('Food & Drug Stores', Decimal('-502.2'), Decimal('1217.4285714285714286'), Decimal('4173'), Decimal('1613.041448851915')),
 ('Apparel', Decimal('396'), Decimal('1263.7000000000000000'), Decimal('3760'), Decimal('1419.134570786013')),
 ('Hotels, Restaurants & Leisure', Decimal('348'), Decimal('1451.0600000000000000'), Decimal('4686.5'), Decimal('1372.975732730432')),
 ('Household Products', Decimal('-198.8'), Decimal('1650.3083333333333333'), Decimal('10508'), Decimal('2900.116805001398')),
 ('Transportation', Decimal('69'), Decimal('1670.2941176470588235'), Decimal('4373'), Decimal('1373.013160657332')),
 ('Industrials', Decimal('-176.1'), Decimal('1727.6894736842105263'), Decimal('8831'), Decimal('2326.018251073599')),
 ('Media', Decimal('-495.9'), Decimal('1821.3363636363636364'), Decimal('9391'), Decimal('2839.299478136369')),
 ('Motor Vehicles & Parts', Decimal('-674.9'), Decimal('1919.5333333333333333'), Decimal('9427'), Decimal('3176.300731983670')),
 ('Aerospace & Defense', Decimal('-941'), Decimal('2093.3083333333333333'), Decimal('5302'), Decimal('2064.779951937795')),
 ('Food, Beverages & Tobacco', Decimal('-677'), Decimal('2346.1833333333333333'), Decimal('14239'), Decimal('3412.352156334481')),
 ('Financials', Decimal('-1128'), Decimal('2719.7761904761904762'), Decimal('24733'), Decimal('5064.764070852874')),
 ('Health Care', Decimal('-1721'), Decimal('2773.2605263157894737'), Decimal('16540'), Decimal('3751.818796086771')),
 ('Telecommunications', Decimal('-383.5'), Decimal('4127.2800000000000000'), Decimal('13127'), Decimal('5400.731732686270')),
 ('Technology', Decimal('-1672'), Decimal('4137.2418604651162791'), Decimal('45687'), Decimal('8042.983363606666'))]

### Summarize group statistics


In [33]:
%%sql

-- Compute standard deviation of maximum values
SELECT STDDEV(max_results.maxval),
	   -- min
       MIN(max_results.maxval),
       -- max
       MAX(max_results.maxval),
       -- avg
       AVG(max_results.maxval)
  -- Subquery to compute max of question_count by tag
  FROM (SELECT tag, MAX(question_count) AS maxval
          FROM stackoverflow
         -- Compute max by...
         GROUP BY tag) AS max_results; -- alias for subquery

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(Decimal('176458.37952720'), 30, 1138658, Decimal('52652.433962264151'))]

## Exploring distributions
### Truncate

In [35]:
%%sql

SELECT trunc(employees, -5) AS employee_bin,
       -- Count number of companies with each truncated value
       COUNT(NAME)
  FROM fortune500
 -- Use alias to group
 GROUP BY employee_bin
 -- Use alias to order
 ORDER BY employee_bin;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
6 rows affected.


[(Decimal('0'), 433),
 (Decimal('100000'), 35),
 (Decimal('200000'), 20),
 (Decimal('300000'), 7),
 (Decimal('400000'), 4),
 (Decimal('2300000'), 1)]

In [36]:
%%sql

SELECT TRUNC(employees, -4) AS employee_bin,
       -- Count number of companies with each truncated value
       COUNT(*)
  FROM fortune500
 -- Limit to which companies?
 WHERE employees < 100000
 -- Use alias to group
 GROUP BY employee_bin
 -- Use alias to order
 ORDER BY employee_bin;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[(Decimal('0'), 102),
 (Decimal('10000'), 108),
 (Decimal('20000'), 63),
 (Decimal('30000'), 42),
 (Decimal('40000'), 35),
 (Decimal('50000'), 31),
 (Decimal('60000'), 18),
 (Decimal('70000'), 18),
 (Decimal('80000'), 6),
 (Decimal('90000'), 10)]

### Generate series


In [37]:
%%sql

-- Select the min and max of question_count
SELECT MIN(question_count), 
       MAX(question_count)
  -- From what table?
  FROM stackoverflow
 -- For tag dropbox
 WHERE tag = 'dropbox';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(2315, 3072)]

In [38]:
%%sql

-- Create lower and upper bounds of bins
SELECT generate_series(2200, 3050, 50) AS lower,
       generate_series(2250, 3100, 50) AS upper;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
18 rows affected.


[(2200, 2250),
 (2250, 2300),
 (2300, 2350),
 (2350, 2400),
 (2400, 2450),
 (2450, 2500),
 (2500, 2550),
 (2550, 2600),
 (2600, 2650),
 (2650, 2700),
 (2700, 2750),
 (2750, 2800),
 (2800, 2850),
 (2850, 2900),
 (2900, 2950),
 (2950, 3000),
 (3000, 3050),
 (3050, 3100)]

In [39]:
%%sql

-- Bins created in Step 2
WITH bins AS (
      SELECT generate_series(2200, 3050, 50) AS lower,
             generate_series(2250, 3100, 50) AS upper),
     -- Subset stackoverflow to just tag dropbox (Step 1)
     dropbox AS (
      SELECT question_count 
        FROM stackoverflow
       WHERE tag='dropbox') 
        

SELECT lower, upper, count(question_count) 
  FROM bins  
       -- Join to dropbox (created above), 
       -- keeping all rows from the bins table in the join
       LEFT JOIN dropbox
       -- Compare question_count to lower and upper
         ON question_count >= lower 
        AND question_count < upper
        
 -- Group by lower and upper to count values in each bin
 GROUP BY lower, upper
 -- Order by lower to put bins in order
 ORDER BY lower;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
18 rows affected.


[(2200, 2250, 0),
 (2250, 2300, 0),
 (2300, 2350, 22),
 (2350, 2400, 39),
 (2400, 2450, 54),
 (2450, 2500, 53),
 (2500, 2550, 45),
 (2550, 2600, 41),
 (2600, 2650, 46),
 (2650, 2700, 57),
 (2700, 2750, 44),
 (2750, 2800, 50),
 (2800, 2850, 62),
 (2850, 2900, 61),
 (2900, 2950, 108),
 (2950, 3000, 159),
 (3000, 3050, 98),
 (3050, 3100, 44)]

## More summary functions
### Correlation


In [40]:
%%sql

-- Correlation between revenues and profit
SELECT CORR(revenues, profits) AS rev_profits,
	   -- Correlation between revenues and assets
       CORR(revenues, assets) AS rev_assets,
       -- Correlation between revenues and equity
       CORR(revenues, equity) AS rev_equity 
  FROM fortune500;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(0.5999935815724782, 0.3294995213185064, 0.546570999718431)]

> Profits, assets, and equity are all positvely correlated with revenue for Fortune 500 companies.

### Mean and Median


In [41]:
%%sql

-- What groups are you computing statistics by?
SELECT sector,
       -- Select the mean of assets with the avg function
       AVG(assets) AS mean,
       -- Select the median
       percentile_disc(.5) WITHIN GROUP (ORDER BY assets) AS median
  FROM fortune500
 -- Computing statistics for each what?
 GROUP BY sector
 -- Order results by a value of interest
 ORDER BY median;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
21 rows affected.


[('Wholesalers', Decimal('9362.5862068965517241'), Decimal('5390')),
 ('Materials', Decimal('10833.263157894737'), Decimal('7741')),
 ('Retailing', Decimal('14473.148936170213'), Decimal('7858')),
 ('Engineering & Construction', Decimal('8199.2307692307692308'), Decimal('8709')),
 ('Apparel', Decimal('11064.8000000000000000'), Decimal('9739')),
 ('Motor Vehicles & Parts', Decimal('59443.111111111111'), Decimal('9901')),
 ('Household Products', Decimal('23179.083333333333'), Decimal('10231')),
 ('Business Services', Decimal('19626.100000000000'), Decimal('12485')),
 ('Hotels, Restaurants & Leisure', Decimal('16795.400000000000'), Decimal('14330')),
 ('Industrials', Decimal('40117.684210526316'), Decimal('15201')),
 ('Chemicals', Decimal('20151.214285714286'), Decimal('15769')),
 ('Food, Beverages & Tobacco', Decimal('29059.750000000000'), Decimal('15984')),
 ('Food & Drug Stores', Decimal('24630.714285714286'), Decimal('17464')),
 ('Technology', Decimal('48380.325581395349'), Decimal('17585')),
 ('Aerospace & Defense', Decimal('31897.666666666667'), Decimal('20038')),
 ('Media', Decimal('30876.272727272727'), Decimal('22508')),
 ('Transportation', Decimal('25922.000000000000'), Decimal('23286')),
 ('Health Care', Decimal('42078.894736842105'), Decimal('25396')),
 ('Telecommunications', Decimal('114922.700000000000'), Decimal('30879')),
 ('Energy', Decimal('48756.210526315789'), Decimal('36119')),
 ('Financials', Decimal('319245.095238095238'), Decimal('123449'))]

In [42]:
%%sql

-- What groups are you computing statistics by?
SELECT sector,
       -- Select the mean of assets with the avg function
       AVG(assets) AS mean,
       -- Select the median
       percentile_cont(.5) WITHIN GROUP (ORDER BY assets) AS median
  FROM fortune500
 -- Computing statistics for each what?
 GROUP BY sector
 -- Order results by a value of interest
 ORDER BY median;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
21 rows affected.


[('Wholesalers', Decimal('9362.5862068965517241'), 5390.0),
 ('Materials', Decimal('10833.263157894737'), 7741.0),
 ('Retailing', Decimal('14473.148936170213'), 7858.0),
 ('Engineering & Construction', Decimal('8199.2307692307692308'), 8709.0),
 ('Apparel', Decimal('11064.8000000000000000'), 9739.0),
 ('Motor Vehicles & Parts', Decimal('59443.111111111111'), 9901.0),
 ('Household Products', Decimal('23179.083333333333'), 11177.0),
 ('Industrials', Decimal('40117.684210526316'), 15201.0),
 ('Business Services', Decimal('19626.100000000000'), 15580.0),
 ('Chemicals', Decimal('20151.214285714286'), 16305.0),
 ('Food, Beverages & Tobacco', Decimal('29059.750000000000'), 16474.5),
 ('Hotels, Restaurants & Leisure', Decimal('16795.400000000000'), 17399.5),
 ('Food & Drug Stores', Decimal('24630.714285714286'), 17464.0),
 ('Technology', Decimal('48380.325581395349'), 17585.0),
 ('Media', Decimal('30876.272727272727'), 22508.0),
 ('Aerospace & Defense', Decimal('31897.666666666667'), 22826.0),
 ('Transportation', Decimal('25922.000000000000'), 23286.0),
 ('Health Care', Decimal('42078.894736842105'), 25491.0),
 ('Energy', Decimal('48756.210526315789'), 36119.0),
 ('Telecommunications', Decimal('114922.700000000000'), 38948.0),
 ('Financials', Decimal('319245.095238095238'), 124520.5)]

## Creating temporary tables
### Create a temp table



In [47]:
%%sql

DROP TABLE IF EXISTS profit80;

-- Create the temporary table
CREATE TEMP TABLE profit80 AS 
  -- Select the two columns you need; alias as needed
    
  SELECT sector, 
         percentile_disc(0.8) WITHIN GROUP (ORDER BY profits) AS pct80
    FROM fortune500
   -- What do you need to group by?
   GROUP BY sector;

    
-- See what you created: select all columns and rows 
-- from the table you created
SELECT * 
  FROM profit80;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
Done.
21 rows affected.
21 rows affected.


[('Aerospace & Defense', Decimal('4895')),
 ('Apparel', Decimal('1074.1')),
 ('Business Services', Decimal('1401')),
 ('Chemicals', Decimal('1500')),
 ('Energy', Decimal('1311')),
 ('Engineering & Construction', Decimal('602.7')),
 ('Financials', Decimal('3014')),
 ('Food & Drug Stores', Decimal('2025.7')),
 ('Food, Beverages & Tobacco', Decimal('6073')),
 ('Health Care', Decimal('4965')),
 ('Hotels, Restaurants & Leisure', Decimal('1670')),
 ('Household Products', Decimal('2166')),
 ('Industrials', Decimal('3695')),
 ('Materials', Decimal('496')),
 ('Media', Decimal('2755')),
 ('Motor Vehicles & Parts', Decimal('4596')),
 ('Retailing', Decimal('1228')),
 ('Technology', Decimal('7266')),
 ('Telecommunications', Decimal('8695')),
 ('Transportation', Decimal('2676')),
 ('Wholesalers', Decimal('605.9'))]

In [46]:
%%sql

-- Select columns, aliasing as needed
SELECT title, fortune500.sector, 
       profits, profits/pct80 AS ratio
-- What tables do you need to join?  
  FROM fortune500 
       LEFT JOIN profit80
-- How are the tables joined?
       ON fortune500.sector=profit80.sector
-- What rows do you want to select?
 WHERE profits > pct80
 LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('Lockheed Martin', 'Aerospace & Defense', Decimal('5302'), Decimal('1.0831460674157303')),
 ('United Technologies', 'Aerospace & Defense', Decimal('5055'), Decimal('1.0326864147088866')),
 ('Nike', 'Apparel', Decimal('3760'), Decimal('3.5006051578065357')),
 ('S&P Global', 'Business Services', Decimal('2106'), Decimal('1.5032119914346895')),
 ('Mastercard', 'Business Services', Decimal('4059'), Decimal('2.8972162740899358')),
 ('ADP', 'Business Services', Decimal('1492.5'), Decimal('1.0653104925053533')),
 ('Visa', 'Business Services', Decimal('5991'), Decimal('4.2762312633832976')),
 ('DuPont', 'Chemicals', Decimal('2513'), Decimal('1.6753333333333333')),
 ('Dow Chemical', 'Chemicals', Decimal('4318'), Decimal('2.8786666666666667')),
 ('PPL', 'Energy', Decimal('1902'), Decimal('1.4508009153318078')),
 ('Sempra Energy', 'Energy', Decimal('1370'), Decimal('1.0450038138825324')),
 ('Dominion Energy', 'Energy', Decimal('2123'), Decimal('1.6193745232646834')),
 ('NextEra Energy', 'Energy', Decimal('2912'), Decimal('2.2212051868802441')),
 ('PG&E Corp.', 'Energy', Decimal('1393'), Decimal('1.0625476735316552')),
 ('Southern', 'Energy', Decimal('2448'), Decimal('1.8672768878718535')),
 ('Enterprise Products Partners', 'Energy', Decimal('2513.1'), Decimal('1.9169336384439359')),
 ('Duke Energy', 'Energy', Decimal('2152'), Decimal('1.6414950419527079')),
 ('Valero Energy', 'Energy', Decimal('2289'), Decimal('1.7459954233409611')),
 ('Phillips 66', 'Energy', Decimal('1555'), Decimal('1.1861174675819985')),
 ('Exxon Mobil', 'Energy', Decimal('7840'), Decimal('5.9801678108314264'))]

### Create a temp table to simplify a query


In [48]:
%%sql

DROP TABLE IF EXISTS startdates;

-- Create temp table syntax
CREATE TEMP TABLE startdates AS
-- Compute the minimum date for each what?
SELECT tag,
       MIN(date) AS mindate
  FROM stackoverflow
 -- What do you need to compute the min date for each tag?
 GROUP BY tag;
 
 -- Look at the table you created
 SELECT * 
   FROM startdates;


 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
Done.
53 rows affected.
53 rows affected.


[('amazon-route53', datetime.date(2016, 1, 1)),
 ('google-spreadsheet', datetime.date(2016, 1, 1)),
 ('dropbox', datetime.date(2016, 1, 1)),
 ('amazon-data-pipeline', datetime.date(2016, 9, 1)),
 ('amazon', datetime.date(2016, 1, 1)),
 ('amazon-sns', datetime.date(2016, 9, 1)),
 ('ios', datetime.date(2016, 1, 1)),
 ('amazon-web-services', datetime.date(2016, 1, 1)),
 ('amazon-cloudsearch', datetime.date(2016, 1, 1)),
 ('amazon-ses', datetime.date(2016, 9, 1)),
 ('cognos', datetime.date(2016, 1, 1)),
 ('amazon-ecs', datetime.date(2016, 9, 1)),
 ('ios9', datetime.date(2016, 1, 1)),
 ('amazon-ebs', datetime.date(2016, 9, 1)),
 ('dropbox-api', datetime.date(2016, 1, 1)),
 ('amazon-elb', datetime.date(2016, 9, 1)),
 ('amazon-emr', datetime.date(2016, 9, 1)),
 ('sql-server', datetime.date(2016, 1, 1)),
 ('amazon-cloudfront', datetime.date(2016, 1, 1)),
 ('amazon-cognito', datetime.date(2016, 9, 1)),
 ('amazon-cloudformation', datetime.date(2016, 9, 1)),
 ('amazon-mws', datetime.date(2016, 9, 1)),
 ('amazon-rds-aurora', datetime.date(2016, 9, 1)),
 ('amazon-dynamodb', datetime.date(2016, 9, 1)),
 ('excel', datetime.date(2016, 1, 1)),
 ('actionscript', datetime.date(2016, 1, 1)),
 ('applepayjs', datetime.date(2017, 3, 18)),
 ('paypal', datetime.date(2016, 1, 1)),
 ('amazon-simpledb', datetime.date(2016, 9, 1)),
 ('mongodb', datetime.date(2016, 1, 1)),
 ('amazon-cloudwatch', datetime.date(2016, 9, 1)),
 ('applepay', datetime.date(2017, 3, 18)),
 ('amazon-kinesis', datetime.date(2016, 9, 1)),
 ('osx', datetime.date(2016, 1, 1)),
 ('amazon-vpc', datetime.date(2016, 9, 1)),
 ('windows', datetime.date(2016, 1, 1)),
 ('amazon-sqs', datetime.date(2016, 9, 1)),
 ('azure', datetime.date(2016, 1, 1)),
 ('citrix', datetime.date(2016, 1, 1)),
 ('amazon-redshift', datetime.date(2016, 9, 1)),
 ('amazon-ec2', datetime.date(2016, 1, 1)),
 ('amazon-glacier', datetime.date(2016, 9, 1)),
 ('actionscript-3', datetime.date(2016, 1, 1)),
 ('amazon-swf', datetime.date(2016, 9, 1)),
 ('stripe-payments', datetime.date(2016, 1, 1)),
 ('amazon-lambda', datetime.date(2016, 9, 1)),
 ('amazon-elastic-beanstalk', datetime.date(2016, 9, 1)),
 ('android-pay', datetime.date(2017, 3, 17)),
 ('ios8', datetime.date(2016, 1, 1)),
 ('amazon-s3', datetime.date(2016, 1, 1)),
 ('amazon-elasticache', datetime.date(2016, 9, 1)),
 ('android', datetime.date(2016, 1, 1)),
 ('amazon-rds', datetime.date(2016, 9, 1))]

In [49]:
%%sql

SELECT startdates.tag, 
       startdates.mindate, 
       -- Select question count on the min and max days
	   so_min.question_count AS min_date_question_count,
       so_max.question_count AS max_date_question_count,
       -- Compute the change in question_count (max- min)
       so_max.question_count - so_min.question_count AS change
  FROM startdates
       -- Join startdates to stackoverflow with alias so_min
       INNER JOIN stackoverflow AS so_min
          -- What needs to match between tables?
          ON startdates.tag = so_min.tag
         AND startdates.mindate = so_min.date
       -- Join to stackoverflow again with alias so_max
       INNER JOIN stackoverflow AS so_max
       	  -- Again, what needs to match between tables?
          ON startdates.tag = so_max.tag
         AND so_max.date = '2018-09-25';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
53 rows affected.


[('paypal', datetime.date(2016, 1, 1), 13296, 18050, 4754),
 ('amazon-elb', datetime.date(2016, 9, 1), 576, 1452, 876),
 ('amazon-mws', datetime.date(2016, 9, 1), 367, 706, 339),
 ('amazon-swf', datetime.date(2016, 9, 1), 167, 232, 65),
 ('amazon-sns', datetime.date(2016, 9, 1), 690, 1400, 710),
 ('excel', datetime.date(2016, 1, 1), 81384, 177603, 96219),
 ('mongodb', datetime.date(2016, 1, 1), 55510, 104159, 48649),
 ('amazon-glacier', datetime.date(2016, 9, 1), 118, 192, 74),
 ('amazon-route53', datetime.date(2016, 1, 1), 369, 1098, 729),
 ('dropbox', datetime.date(2016, 1, 1), 2319, 3071, 752),
 ('azure', datetime.date(2016, 1, 1), 25859, 61259, 35400),
 ('sql-server', datetime.date(2016, 1, 1), 151267, 242484, 91217),
 ('amazon-cloudwatch', datetime.date(2016, 9, 1), 334, 1165, 831),
 ('amazon-redshift', datetime.date(2016, 9, 1), 1448, 3502, 2054),
 ('amazon-cognito', datetime.date(2016, 9, 1), 547, 2830, 2283),
 ('amazon-dynamodb', datetime.date(2016, 9, 1), 2687, 5985, 3298),
 ('ios9', datetime.date(2016, 1, 1), 3147, 4183, 1036),
 ('amazon-sqs', datetime.date(2016, 9, 1), 855, 1584, 729),
 ('dropbox-api', datetime.date(2016, 1, 1), 1437, 2072, 635),
 ('amazon-ses', datetime.date(2016, 9, 1), 481, 934, 453),
 ('amazon-emr', datetime.date(2016, 9, 1), 557, 3046, 2489),
 ('citrix', datetime.date(2016, 1, 1), 465, 597, 132),
 ('amazon-elastic-beanstalk', datetime.date(2016, 9, 1), 281, 5490, 5209),
 ('amazon-rds', datetime.date(2016, 9, 1), 1156, 2537, 1381),
 ('amazon-rds-aurora', datetime.date(2016, 9, 1), 60, 323, 263),
 ('android', datetime.date(2016, 1, 1), 770606, 1138658, 368052),
 ('amazon-cloudformation', datetime.date(2016, 9, 1), 592, 2516, 1924),
 ('ios', datetime.date(2016, 1, 1), 397485, 577421, 179936),
 ('android-pay', datetime.date(2017, 3, 17), 444, 490, 46),
 ('applepay', datetime.date(2017, 3, 18), 222, 357, 135),
 ('windows', datetime.date(2016, 1, 1), 88874, 124751, 35877),
 ('amazon-web-services', datetime.date(2016, 1, 1), 22077, 63939, 41862),
 ('google-spreadsheet', datetime.date(2016, 1, 1), 6531, 15685, 9154),
 ('amazon-cloudfront', datetime.date(2016, 1, 1), 1099, 2264, 1165),
 ('amazon-lambda', datetime.date(2016, 9, 1), 158, 8032, 7874),
 ('amazon-cloudsearch', datetime.date(2016, 1, 1), 176, 318, 142),
 ('amazon-ecs', datetime.date(2016, 9, 1), 145, 1074, 929),
 ('amazon-s3', datetime.date(2016, 1, 1), 12015, 25443, 13428),
 ('amazon-elasticache', datetime.date(2016, 9, 1), 257, 434, 177),
 ('amazon-simpledb', datetime.date(2016, 9, 1), 374, 392, 18),
 ('amazon-data-pipeline', datetime.date(2016, 9, 1), 195, 399, 204),
 ('amazon-ec2', datetime.date(2016, 1, 1), 11632, 22755, 11123),
 ('amazon-ebs', datetime.date(2016, 9, 1), 257, 338, 81),
 ('ios8', datetime.date(2016, 1, 1), 9854, 9399, -455),
 ('amazon-kinesis', datetime.date(2016, 9, 1), 259, 766, 507),
 ('cognos', datetime.date(2016, 1, 1), 616, 848, 232),
 ('actionscript', datetime.date(2016, 1, 1), 8761, 9101, 340),
 ('amazon', datetime.date(2016, 1, 1), 2994, 3636, 642),
 ('osx', datetime.date(2016, 1, 1), 62614, 87485, 24871),
 ('stripe-payments', datetime.date(2016, 1, 1), 1736, 4964, 3228),
 ('actionscript-3', datetime.date(2016, 1, 1), 39025, 41149, 2124),
 ('amazon-vpc', datetime.date(2016, 9, 1), 339, 719, 380),
 ('applepayjs', datetime.date(2017, 3, 18), 11, 30, 19)]

### Insert into a temp table


In [50]:
%%sql

DROP TABLE IF EXISTS correlations;

-- Create temp table 
CREATE TEMP TABLE correlations AS
-- Select each correlation
SELECT 'profits'::varchar AS measure,
       -- Compute correlations
       CORR(profits, profits) AS profits,
       CORR(profits, profits_change) AS profits_change,
       CORR(profits, revenues_change) AS revenues_change
  FROM fortune500;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
Done.
1 rows affected.


[]

In [51]:
%%sql

INSERT INTO correlations
-- Follow the pattern of the select statement above
-- Using profits_change instead of profits
SELECT 'profits_change'::varchar AS measure,
       corr(profits_change, profits) AS profits,
       corr(profits_change, profits_change) AS profits_change,
       corr(profits_change, revenues_change) AS revenues_change
  FROM fortune500;

-- Repeat the above, but for revenues_change

INSERT INTO correlations
SELECT 'revenues_change'::varchar AS measure,
       corr(revenues_change, profits) AS profits,
       corr(revenues_change, profits_change) AS profits_change,
       corr(revenues_change, revenues_change) AS revenues_change
  FROM fortune500;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.
1 rows affected.


[]

In [52]:
%%sql
-- Select each column, rounding the correlations
SELECT measure, 
       round(profits::numeric, 2) AS profits,
       round(profits_change::numeric, 2) AS profits_change,
       round(revenues_change::numeric, 2) AS revenues_change
  FROM correlations;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
3 rows affected.


[('profits', Decimal('1.00'), Decimal('0.02'), Decimal('0.02')),
 ('profits_change', Decimal('0.02'), Decimal('1.00'), Decimal('-0.09')),
 ('revenues_change', Decimal('0.02'), Decimal('-0.09'), Decimal('1.00'))]

# Exploring categorical data and unstructured text

## Character data types and common issues
### Count of the categories

In [53]:
%%sql

-- Select the count of each level of priority
SELECT priority, COUNT(*)
FROM evanston311
GROUP BY priority;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
4 rows affected.


[('MEDIUM', 5745), ('NONE', 30081), ('HIGH', 88), ('LOW', 517)]

In [54]:
%%sql

-- Find values of zip that appear in at least 100 rows
-- Also get the count of each value
SELECT zip, COUNT(*)
FROM evanston311
GROUP BY zip
HAVING COUNT(*) >100; 

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
4 rows affected.


[('60208', 255), (None, 5528), ('60201', 19054), ('60202', 11165)]

In [55]:
%%sql

-- Find values of source that appear in at least 100 rows
-- Also get the count of each value
SELECT source, COUNT(*)
FROM evanston311
GROUP BY source
HAVING COUNT(*) >= 100;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
4 rows affected.


[('gov.publicstuff.com', 30985),
 ('Android', 444),
 ('Iframe', 3670),
 ('iOS', 1199)]

In [56]:
%%sql

-- Find the 5 most common values of street and the count of each
SELECT street, COUNT(*)
FROM evanston311
 GROUP BY street
 ORDER BY count(*) DESC
 LIMIT 5;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
5 rows affected.


[(None, 1699),
 ('Chicago Avenue', 1440),
 ('Sherman Avenue', 1276),
 ('Central Street', 1211),
 ('Davis Street', 1154)]

## Cases and spaces
### Trimming


In [58]:
%%sql

SELECT distinct street,
       -- Trim off unwanted characters from street
       TRIM(street, '0123456789 #/.') AS cleaned_street
  FROM evanston311
 ORDER BY street
LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('1/2 Chicago Ave', 'Chicago Ave'),
 ('1047B Chicago Ave', 'B Chicago Ave'),
 ('13th Street', 'th Street'),
 ('141A Callan Ave', 'A Callan Ave'),
 ('141b Callan Ave', 'b Callan Ave'),
 ('1624B Central St', 'B Central St'),
 ('217A Dodge Ave', 'A Dodge Ave'),
 ('221c Dodge Ave', 'c Dodge Ave'),
 ('300c Dodge Ave', 'c Dodge Ave'),
 ('3314A Central St', 'A Central St'),
 ('36th Street', 'th Street'),
 ('600A South Blvd', 'A South Blvd'),
 ('606B South Blvd', 'B South Blvd'),
 ('612C South Blvd', 'C South Blvd'),
 ('613B Custer Ave', 'B Custer Ave'),
 ('618B South Blvd', 'B South Blvd'),
 ('6th Street', 'th Street'),
 ('Arlington Boulevard', 'Arlington Boulevard'),
 ('Arnold Pl', 'Arnold Pl'),
 ('Arts Cir Dr', 'Arts Cir Dr')]

### Exploring unstructured text


In [60]:
%%sql

SELECT COUNT(*)
  FROM evanston311
 -- Where description includes trash or garbage
 WHERE description ILIKE '%trash%'
    OR description ILIKE '%garbage%';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(2551,)]

In [61]:
%%sql

-- Select categories containing Trash or Garbage
SELECT category
  FROM evanston311
 -- Use LIKE
 WHERE category LIKE '%Trash%'
    OR category LIKE '%Garbage%'
LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('THIS REQUEST IS INACTIVE...Trash Cart - Compost Bin',),
 ('Trash - Tire Pickup',),
 ('Trash - Special Pickup - Resident Use',),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement',),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement',),
 ('Trash - Missed Garbage Pickup',),
 ('THIS REQUEST IS INACTIVE...Trash Cart - Compost Bin',),
 ('Trash - Tire Pickup',),
 ('Trash - Missed Garbage Pickup',),
 ('Trash - Accumulation',),
 ('Trash - Special Pickup (STAFF ONLY)',),
 ('THIS REQUEST IS INACTIVE...Trash Cart - Compost Bin',),
 ('Trash Cart, Recycling Cart - Missing',),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement',),
 ('Trash - Accumulation',),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement',),
 ('Trash - Missed Garbage Pickup',),
 ('Trash - Garbage in Parks or Right of Way',),
 ('Trash - Missed Garbage Pickup',),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement',)]

In [63]:
%%sql

SELECT COUNT(*)
  FROM evanston311 
 -- description contains trash or garbage (any case)
 WHERE (description ILIKE '%trash%'
    OR description ILIKE '%garbage%') 
 -- category does not contain Trash or Garbage
   AND category NOT LIKE '%Trash%'
   AND category NOT LIKE '%Garbage%';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(570,)]

In [64]:
%%sql

-- Count rows with each category
SELECT category, COUNT(*)
  FROM evanston311 
 WHERE (description ILIKE '%trash%'
    OR description ILIKE '%garbage%') 
   AND category NOT LIKE '%Trash%'
   AND category NOT LIKE '%Garbage%'
 -- What are you counting?
 GROUP BY category
 --- order by most frequent values
 ORDER BY count DESC
 LIMIT 10;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
10 rows affected.


[('Ask A Question / Send A Message', 273),
 ('Rodents- Rats', 77),
 ('Recycling - Missed Pickup', 28),
 ('Dead Animal on Public Property', 16),
 ('Graffiti', 15),
 ('Yard Waste - Missed Pickup', 14),
 ('Public Transit Agency Issue', 13),
 ('Food Establishment - Unsanitary Conditions', 13),
 ('Exterior Conditions', 10),
 ('Street Sweeping', 9)]

> The results include some categories that appear to be related to trash, while others are more general.

## Splitting and concatenating text
### Concatenate strings



In [69]:
%%sql

-- Concatenate house_num, a space, and street
-- and trim spaces from the start of the result
SELECT ltrim(concat(house_num, ' ', street)) AS address
  FROM evanston311
LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('606-612 Sheridan Road',),
 ('930 Washington St',),
 ('1183-1223 Lincoln St',),
 ('1–111 Callan Ave',),
 ('1524 Crain St',),
 ('2830 Central Street',),
 ('1139 Dodge Ave',),
 ('900 Oakton Street',),
 ('608 Oakton Street',),
 ('1320 Dewey Avenue',),
 ('815 Mulford St',),
 ('1714 Greenwood Street',),
 ('148 Clyde Avenue',),
 ('1848 Ashland Avenue',),
 ('2734 Garrison Avenue',),
 ('1030 Dodge Ave',),
 ('',),
 ('1720 Brummel St',),
 ('1600 Sherman Avenue',),
 ('2308 Foster St',)]

### Split strings on a delimiter


In [70]:
%%sql

-- Select the first word of the street value
SELECT split_part(street, ' ', 1) AS street_name, 
       count(*)
  FROM evanston311
 GROUP BY street_name
 ORDER BY count DESC
 LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[(None, 1699),
 ('Chicago', 1569),
 ('Central', 1529),
 ('Sherman', 1479),
 ('Davis', 1248),
 ('Church', 1225),
 ('Main', 880),
 ('Sheridan', 842),
 ('Ridge', 823),
 ('Dodge', 816),
 ('Maple', 778),
 ('Asbury', 675),
 ('Hinman', 586),
 ('West', 578),
 ('Orrington', 561),
 ('Emerson', 513),
 ('Grove', 498),
 ('Darrow', 489),
 ('Custer', 464),
 ('Lake', 444)]

### Shorten long strings


In [72]:
%%sql

-- Select the first 50 chars when length is greater than 50
SELECT CASE WHEN length(description) > 50
            THEN left(description, 50) || '...'
       -- otherwise just select description
       ELSE description
       END
  FROM evanston311
 -- limit to descriptions that start with the word I
 WHERE description LIKE 'I %'
 ORDER BY description
 LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('I  work for Schermerhorn & Co. and manage this con...',),
 ('I Live in a townhouse with garbage cans in back, i...',),
 ('I Put In For Reserve Disabled Parking, A Week Ago ...',),
 ('I SDO GOWANS #1258 RECEIVED A TELEPHONE CALL ON 3/...',),
 ('I accidentally mistyped my license plate number - ...',),
 ('I accidentally sent the wrong cover letter on my a...',),
 ('I acquired c diff at north shore hospital in Evans...',),
 ('I am a 35 year resident of Evanston (314 Custer Av...',),
 ('I am a Cubs fan and watched game seven. But using ...',),
 ('I am a Northwestern student that has accumulated t...',),
 ('I am a PTA president at Dewey Elementary.  Alongsi...',),
 ('I am a business owner at 1121 Emerson St at the co...',),
 ('I am a current customer at 1333 Maple Ave, Unit 2E...',),
 ('I am a day care worker at the family center at the...',),
 ('I am a senior citizen and need to dispose of elect...',),
 ('I am about to renew my Pooch Park pass for 2 dogs ...',),
 ('I am an Evanston resident.  I would like to disput...',),
 ('I am aware that the city of Evanston doesn’t regul...',),
 ('I am concerned about the relatively young parkway ...',),
 ('I am contacting you from Rimland Services. We have...',)]

## Strategies for multiple transformations
### Create an "other" category


In [73]:
%%sql

SELECT CASE WHEN zipcount < 100 THEN 'other'
       ELSE zip
       END AS zip_recoded,
       sum(zipcount) AS zipsum
  FROM (SELECT zip, count(*) AS zipcount
          FROM evanston311
         GROUP BY zip) AS fullcounts
 GROUP BY zip_recoded
 ORDER BY zipsum DESC;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
5 rows affected.


[('60201', Decimal('19054')),
 ('60202', Decimal('11165')),
 (None, Decimal('5528')),
 ('other', Decimal('429')),
 ('60208', Decimal('255'))]

> All of the zip codes with fewer observations than 60208 have less than 100 rows in the table.

### Group and recode values


In [79]:
%%sql

-- Fill in the command below with the name of the temp table
DROP TABLE IF EXISTS recode;

-- Create and name the temporary table
CREATE TEMP TABLE recode AS
-- Write the select query to generate the table 
-- with distinct values of category and standardized values
  SELECT DISTINCT category, 
         rtrim(split_part(category, '-', 1)) AS standardized
    -- What table are you selecting the above values from?
    FROM evanston311;
    
-- Look at a few values before the next step
SELECT DISTINCT standardized 
  FROM recode
 WHERE standardized LIKE 'Trash%Cart'
    OR standardized LIKE 'Snow%Removal%';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
Done.
149 rows affected.
6 rows affected.


[('Snow Removal',),
 ('Snow Removal/Concerns',),
 ('Snow/Ice/Hazard Removal',),
 ('Trash Cart',),
 ('Trash Cart, Recycling Cart',),
 ('Trash, Recycling, Yard Waste Cart',)]

In [80]:
%%sql

-- Update to group trash cart values
UPDATE recode 
   SET standardized='Trash Cart' 
 WHERE standardized LIKE 'Trash%Cart';

-- Update to group snow removal values
UPDATE recode 
   SET standardized='Snow Removal' 
 WHERE standardized LIKE 'Snow%Removal%';
    
-- Examine effect of updates
SELECT DISTINCT standardized 
  FROM recode
 WHERE standardized LIKE 'Trash%Cart'
    OR standardized LIKE 'Snow%Removal%';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
3 rows affected.
5 rows affected.
2 rows affected.


[('Snow Removal',), ('Trash Cart',)]

In [81]:
%%sql

UPDATE recode SET standardized='UNUSED' 
 WHERE standardized IN ('THIS REQUEST IS INACTIVE...Trash Cart', 
               '(DO NOT USE) Water Bill',
               'DO NOT USE Trash', 'NO LONGER IN USE');

-- Select the recoded categories and the count of each
SELECT standardized, count(*) 
-- From the original table and table with recoded values
  FROM evanston311 
       LEFT JOIN recode 
       -- What column do they have in common?
       ON evanston311.category=recode.category 
 -- What do you need to group by to count?
 GROUP BY standardized
 -- Display the most common val values first
 ORDER BY count DESC;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
4 rows affected.
115 rows affected.


[('Broken Parking Meter', 6092),
 ('Trash', 3699),
 ('Ask A Question / Send A Message', 2595),
 ('Trash Cart', 1902),
 ('Tree Evaluation', 1879),
 ('Rodents', 1305),
 ('Recycling', 1224),
 ('Dead Animal on Public Property', 1057),
 ('Child Seat Installation or Inspection', 1028),
 ('Fire Prevention', 880),
 ('Water Bill', 878),
 ('Pot Hole', 849),
 ('Yard Waste', 811),
 ('Broken Pay Station', 799),
 ('Trees', 739),
 ('UNUSED', 679),
 ('Street Signs', 656),
 ('Sidewalk', 641),
 ('Graffiti', 638),
 ('Water Bill Reprint, Request an Address Change and/or Name Change', 615),
 ('Traffic Signal/Traffic Signal Back Plate', 569),
 ('Street Light', 487),
 ('Forestry', 427),
 ('Sewer', 393),
 ('Street Lights', 365),
 ('Rental Dwelling Inspection or Overcrowding', 300),
 ('Overgrown Grass or Weeds', 271),
 ('Water Service', 246),
 ('Construction/Engineering', 206),
 ('Snow Removal', 195),
 ('Dead Rodent', 192),
 ('Broken Park/Playground Equipment', 190),
 ('Alarm Registration', 178),
 ('Street Sweeping', 169),
 ('Leaf Blower Complaint', 167),
 ('Water Main Break/Service Leak', 160),
 ('Park Lawn Maintenance', 160),
 ('Food Establishment', 152),
 ('Illegal Construction', 144),
 ('Bathroom Maintenance/City Parks', 135),
 ('Abandoned Bicycle on City Property', 132),
 ('Illegally Posted Sign(s) on Public or Private Property', 128),
 ('Exterior Conditions', 114),
 ('Health/Miscellaneous', 112),
 ('Water Meter', 111),
 ('Garbage', 108),
 ('Private Utility Service Issue', 100),
 ('Parkway Restoration Following Water, Sewer, Repair', 99),
 ('Repair Park or Play Lot Fountain', 92),
 ('Park Light(s)', 87),
 ('Power Outage', 85),
 ('Open/Noisy Manhole Cover on a City Street or Parkway', 77),
 ('Birds', 73),
 ('Public Transit Agency Issue', 71),
 ('Bee/Wasp Infestation', 70),
 ('Broken Fence or Bench', 67),
 ('Animals', 59),
 ('Low Water Pressure', 58),
 ('Tree/Branch Debris in Alley', 57),
 ('Water Quality', 45),
 ('Speed Bumps/Speed Humps', 38),
 ('Broken Buffalo Box', 38),
 ('Check my Wheel Tax Payment Status?', 31),
 ('Advanced Disposal', 31),
 ('Elevator Concern', 30),
 ('Amplified Sounds and/or Music', 29),
 ('Insects', 29),
 ('Bike Racks', 27),
 ('Garage, Yard, or Estate Sale Permit', 26),
 ('General/Routine Maintenance', 26),
 ('Fire Hydrant', 23),
 ('Smoking', 21),
 ('Dropped Keys/item in Drainage Structure', 20),
 ('Collection Boxes', 19),
 ('Mosquitoes', 18),
 ('Bike Routes/Bike Lanes Road Maintenance', 18),
 ('Water Service Disruption', 15),
 ('Heat Issue', 14),
 ('Bats', 13),
 ('Property Zoning Information', 9),
 ('Sanitation Billing Questions', 9),
 ('Notice of Violation', 9),
 ('Paint Markings', 9),
 ('Plastic Bag Ordinance', 8),
 ('Water Plant Tour', 7),
 ('Electric Aggregation/Solicitor Issues', 7),
 ('Parking App', 6),
 ('Taxi Complaints', 6),
 ('Accessibility', 5),
 ('Christmas Tree Pick', 5),
 ('Park Programs Information', 5),
 ('Animal Issue/Concern', 5),
 ('Police Open House (Questions)', 4),
 ('Abandoned Vehicle', 4),
 ('Renting an athletic field(s)', 4),
 ('Recycling Education Information/Materials', 4),
 ('Alleys', 3),
 ('Post Office', 3),
 ('Newspaper Box Publication Display Permit Application', 3),
 ('Landmark Building Plaque or Nomination', 3),
 ('Lawn Irrigation Violation', 3),
 ('Refuse Cart Placement', 3),
 ('ADA/Inclusion Aids', 3),
 ('Food Waste', 3),
 ('Illegal Trespass at a Vacant Building', 3),
 ('Move Out', 2),
 ('Animal Service', 2),
 ('Long Term Care', 2),
 ('Temporary No Parking Signs', 2),
 ('Rental Registration Question/Concern', 2),
 ('General Payment Question', 1),
 ('Line Down', 1),
 ('Illegally Placed Newsrack', 1),
 ('Key Request', 1),
 ('Food Waste, Collective Resource', 1)]

### Create a table with indicator variables


In [101]:
%%sql

-- To clear table if it already exists
DROP TABLE IF EXISTS indicators;

-- Create the indicators temp table
CREATE TEMP TABLE indicators AS
  -- Select id
  SELECT id, 
         -- Create the email indicator (find @)
         CAST (description LIKE '%@%' AS integer) AS email,
         -- Create the phone indicator
         CAST (description LIKE '%___-___-____%' AS integer) AS phone 
    -- What table contains the data? 
    FROM evanston311;

-- Inspect the contents of the new temp table
SELECT *
  FROM indicators
  LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
Done.
36431 rows affected.
20 rows affected.


[(1340563, 0, 0),
 (1826017, 0, 0),
 (1849204, 0, 0),
 (1880254, 0, 0),
 (1972582, 0, 1),
 (1840025, 0, 0),
 (2099219, 0, 0),
 (2554820, None, None),
 (1770749, 0, 0),
 (2129641, 0, 1),
 (2167024, 1, 1),
 (2181980, 0, 0),
 (2198222, 0, 0),
 (2561782, 0, 0),
 (2662755, 0, 0),
 (2195772, 0, 0),
 (2250632, 0, 0),
 (2406453, 0, 0),
 (2523257, None, None),
 (2552716, 0, 0)]

In [102]:
%%sql

-- Select the column you'll group by
SELECT priority,
       -- Compute the proportion of rows with each indicator
       SUM(email)/count(*)::numeric AS email_prop, 
       SUM(phone)/count(*)::numeric AS phone_prop
  -- Tables to select from
  FROM evanston311
       LEFT JOIN indicators
       -- Joining condition
       ON evanston311.id=indicators.id
 -- What are you grouping by?
 GROUP BY priority;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
4 rows affected.


[('MEDIUM', Decimal('0.01966927763272410792'), Decimal('0.01845082680591818973')),
 ('NONE', Decimal('0.00412220338419600412'), Decimal('0.00568465144110900568')),
 ('HIGH', Decimal('0.01136363636363636364'), Decimal('0.02272727272727272727')),
 ('LOW', Decimal('0.00580270793036750484'), Decimal('0.00193423597678916828'))]

# Working with dates and timestamps
## Date/time types and formats
### Date comparisons


In [103]:
%%sql

-- Count requests created on January 31, 2017
SELECT count(*) 
  FROM evanston311
 WHERE date_created::date = '2017-01-31';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(45,)]

In [104]:
%%sql

-- Count requests created on February 29, 2016
SELECT count(*)
  FROM evanston311 
 WHERE date_created >= '2016-02-29'
   AND date_created < '2016-03-01';

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(58,)]

In [105]:
%%sql

-- Count requests created on March 13, 2017
SELECT count(*)
  FROM evanston311
 WHERE date_created >= '2017-03-13'
   AND date_created < '2017-03-13'::date + 1;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(33,)]

### Date arithmetic


In [106]:
%%sql

-- Subtract the min date_created from the max
SELECT max(date_created)-min(date_created) 
  FROM evanston311;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(datetime.timedelta(days=911, seconds=59619),)]

In [107]:
%%sql

-- How old is the most recent request?
SELECT now() - max(date_created)
  FROM evanston311;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(datetime.timedelta(days=1473, seconds=22784, microseconds=172917),)]

In [108]:
%%sql

-- Add 100 days to the current timestamp
SELECT NOW() + '100 days'::interval ;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(datetime.datetime(2022, 10, 20, 18, 56, 27, 280054, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))),)]

In [109]:
%%sql

-- Select the current timestamp, 
-- and the current timestamp + 5 minutes
SELECT NOW() + '5 minutes'::interval;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(datetime.datetime(2022, 7, 12, 19, 1, 39, 805601, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))),)]

### Completion time by category


In [110]:
%%sql

-- Select the category and the average completion time by category
SELECT category, 
       AVG(date_completed - date_created) AS completion_time
  FROM evanston311
 GROUP BY category
-- Order the results
 ORDER BY completion_time DESC;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
149 rows affected.


[('Rodents- Rats', datetime.timedelta(days=64, seconds=39503, microseconds=766)),
 ('Fire Prevention - Public Education', datetime.timedelta(days=34, seconds=60490)),
 ('Key Request - All  City Employees', datetime.timedelta(days=32, seconds=3131)),
 ('Smoking', datetime.timedelta(days=27, seconds=27750, microseconds=238095)),
 ('Notice of Violation', datetime.timedelta(days=24, seconds=20472, microseconds=666667)),
 ('Exterior Conditions', datetime.timedelta(days=23, seconds=80311, microseconds=87719)),
 ('General/Routine Maintenance - Facilities Management', datetime.timedelta(days=23, seconds=10916, microseconds=480000)),
 ('Trash - Special Pickup (STAFF ONLY)', datetime.timedelta(days=19, seconds=21892, microseconds=96070)),
 ('Public Transit Agency Issue', datetime.timedelta(days=19, seconds=12495, microseconds=845070)),
 ('Private Utility Service Issue', datetime.timedelta(days=17, seconds=71413, microseconds=130000)),
 ('Accessibility', datetime.timedelta(days=15, seconds=6953, microseconds=200000)),
 ('Trash Cart - Downsize, Upsize or Remove', datetime.timedelta(days=13, seconds=78598, microseconds=179842)),
 ('Fire Prevention - Fire Drill', datetime.timedelta(days=13, seconds=55582, microseconds=666667)),
 ('Park Lawn Maintenance', datetime.timedelta(days=13, seconds=31159, microseconds=456250)),
 ('THIS REQUEST IS INACTIVE...Trash Cart - Compost Bin', datetime.timedelta(days=13, seconds=27242, microseconds=845238)),
 ('Trash - Special Pickup - Resident Use', datetime.timedelta(days=12, seconds=74950, microseconds=426606)),
 ('Park Light(s)', datetime.timedelta(days=12, seconds=69297, microseconds=402299)),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement', datetime.timedelta(days=12, seconds=51451, microseconds=307381)),
 ('Sanitation Billing Questions', datetime.timedelta(days=12, seconds=40405, microseconds=888889)),
 ('Repair Park or Play Lot Fountain', datetime.timedelta(days=12, seconds=30758, microseconds=913043)),
 ('Landmark Building Plaque or Nomination', datetime.timedelta(days=11, seconds=18671, microseconds=666667)),
 ('Health/Miscellaneous', datetime.timedelta(days=11, seconds=16322, microseconds=625000)),
 ('Broken Buffalo Box', datetime.timedelta(days=10, seconds=76009, microseconds=289473)),
 ('Insects', datetime.timedelta(days=10, seconds=72803, microseconds=689655)),
 ('Recycling - Missed Pickup', datetime.timedelta(days=10, seconds=49959, microseconds=82517)),
 ('Move Out- Special Pickup', datetime.timedelta(days=10, seconds=28620, microseconds=500000)),
 ('Broken Fence or Bench', datetime.timedelta(days=10, seconds=7339, microseconds=626865)),
 ('Tree Evaluation', datetime.timedelta(days=10, seconds=1136, microseconds=637041)),
 ('Long Term Care', datetime.timedelta(days=9, seconds=83529)),
 ('Forestry- Ask A Question / Send A Message', datetime.timedelta(days=9, seconds=20179, microseconds=398127)),
 ('Broken Park/Playground Equipment', datetime.timedelta(days=9, seconds=19959, microseconds=405263)),
 ('Ask A Question / Send A Message', datetime.timedelta(days=8, seconds=70173, microseconds=538728)),
 ('Construction/Engineering', datetime.timedelta(days=8, seconds=65591, microseconds=344660)),
 ('Rental Dwelling Inspection or Overcrowding', datetime.timedelta(days=8, seconds=64509, microseconds=506667)),
 ('Paint Markings - Question or Concern', datetime.timedelta(days=8, seconds=48848, microseconds=444444)),
 ('Bike Racks', datetime.timedelta(days=8, seconds=41995, microseconds=740741)),
 ('Overgrown Grass or Weeds', datetime.timedelta(days=8, seconds=36771, microseconds=369004)),
 ('Sidewalk- Evaluation', datetime.timedelta(days=8, seconds=1948, microseconds=694051)),
 ('Sidewalk - Maintenance', datetime.timedelta(days=7, seconds=49508, microseconds=649306)),
 ('Garbage - Overflow, odorous', datetime.timedelta(days=7, seconds=41195, microseconds=138889)),
 ('Trash - Accumulation', datetime.timedelta(days=7, seconds=17656, microseconds=483740)),
 ('Abandoned Bicycle on City Property', da

## Date/time components and aggregation
### Date parts

### How many requests are created in each of the 12 months during 2016-2017?

In [111]:
%%sql

-- Extract the month from date_created and count requests
SELECT date_part('month', date_created) AS month, 
       COUNT(*)
  FROM evanston311
 -- Limit the date range
 WHERE date_created >= '2016-01-01'
   AND date_created <= '2018-01-01'
 -- Group by what to get monthly counts?
 GROUP BY month;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
12 rows affected.


[(6.0, 3404),
 (8.0, 3111),
 (4.0, 2385),
 (3.0, 2171),
 (5.0, 2674),
 (10.0, 2399),
 (11.0, 2284),
 (9.0, 2758),
 (12.0, 1998),
 (1.0, 1809),
 (2.0, 1774),
 (7.0, 3063)]

### What is the most common hour of the day for requests to be created?

In [112]:
%%sql

-- Get the hour and count requests
SELECT date_part('hour', date_created) AS hour,
       count(*)
  FROM evanston311
 GROUP BY hour
 -- Order results to select most common
 ORDER BY count DESC
 LIMIT 1;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(5.0, 4089)]

### During what hours are requests usually completed? Count requests completed by hour.

In [113]:
%%sql

-- Count requests completed by hour
SELECT EXTRACT(HOUR FROM date_completed) AS hour,
       COUNT(*)
  FROM evanston311
 GROUP BY hour
 ORDER BY hour;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
24 rows affected.


[(Decimal('0'), 9),
 (Decimal('1'), 50),
 (Decimal('2'), 1296),
 (Decimal('3'), 1870),
 (Decimal('4'), 2516),
 (Decimal('5'), 2744),
 (Decimal('6'), 3162),
 (Decimal('7'), 3351),
 (Decimal('8'), 3580),
 (Decimal('9'), 4787),
 (Decimal('10'), 5059),
 (Decimal('11'), 5242),
 (Decimal('12'), 1605),
 (Decimal('13'), 428),
 (Decimal('14'), 204),
 (Decimal('15'), 131),
 (Decimal('16'), 101),
 (Decimal('17'), 124),
 (Decimal('18'), 69),
 (Decimal('19'), 50),
 (Decimal('20'), 13),
 (Decimal('21'), 10),
 (Decimal('22'), 11),
 (Decimal('23'), 19)]

### Variation by day of week


In [114]:
%%sql

-- Select name of the day of the week the request was created 
SELECT to_char(date_created, 'day') AS day, 
       -- Select avg time between request creation and completion
       AVG(date_completed - date_created) AS duration
  FROM evanston311 
 -- Group by the name of the day of the week and 
 -- integer value of day of week the request was created
 GROUP BY day, EXTRACT(DOW FROM date_created)
 -- Order by integer value of the day of the week 
 -- the request was created
 ORDER BY EXTRACT(DOW FROM date_created);

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
7 rows affected.


[('sunday   ', datetime.timedelta(days=9, seconds=1524, microseconds=787500)),
 ('monday   ', datetime.timedelta(days=7, seconds=2651, microseconds=902579)),
 ('tuesday  ', datetime.timedelta(days=7, seconds=14390, microseconds=799258)),
 ('wednesday', datetime.timedelta(days=7, seconds=39250, microseconds=929181)),
 ('thursday ', datetime.timedelta(days=7, seconds=38945, microseconds=763338)),
 ('friday   ', datetime.timedelta(days=8, seconds=38313, microseconds=994587)),
 ('saturday ', datetime.timedelta(days=7, seconds=55035, microseconds=353755))]

### Date truncation


In [115]:
%%sql

-- Aggregate daily counts by month
SELECT date_trunc('month', day) AS month,
       AVG(count)
  -- Subquery to compute daily counts
  FROM (SELECT date_trunc('day', date_created) AS day,
               COUNT(*) AS count
          FROM evanston311
         GROUP BY day) AS daily_count
 GROUP BY month
 ORDER BY month;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
31 rows affected.


[(datetime.datetime(2015, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('2.0000000000000000')),
 (datetime.datetime(2016, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('23.4516129032258065')),
 (datetime.datetime(2016, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('30.7241379310344828')),
 (datetime.datetime(2016, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.5483870967741935')),
 (datetime.datetime(2016, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('37.3000000000000000')),
 (datetime.datetime(2016, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('40.7666666666666667')),
 (datetime.datetime(2016, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('44.0000000000000000')),
 (datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('41.4838709677419355')),
 (datetime.datetime(2016, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('46.6129032258064516')),
 (datetime.datetime(2016, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('47.2666666666666667')),
 (datetime.datetime(2016, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.8064516129032258')),
 (datetime.datetime(2016, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('36.6551724137931034')),
 (datetime.datetime(2016, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('29.3548387096774194')),
 (datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('34.9032258064516129')),
 (datetime.datetime(2017, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('32.7037037037037037')),
 (datetime.datetime(2017, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.6333333333333333')),
 (datetime.datetime(2017, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('42.2000000000000000')),
 (datetime.datetime(2017, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('46.8064516129032258')),
 (datetime.datetime(2017, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('69.4666666666666667')),
 (datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('57.3225806451612903')),
 (datetime.datetime(2017, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('53.7419354838709677')),
 (datetime.datetime(2017, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('44.6666666666666667')),
 (datetime.datetime(2017, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('41.5806451612903226')),
 (datetime.datetime(2017, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('40.7000000000000000')),
 (datetime.datetime(2017, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.0967741935483871')),
 (datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.4827586206896552')),
 (datetime.datetime(2018, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('30.5714285714285714')),
 (datetime.datetime(2018, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('29.6774193548387097')),
 (datetime.datetim

## Aggregating with date/time series
### Find missing dates


In [116]:
%%sql

SELECT day
-- 1) Subquery to generate all dates
-- from min to max date_created
  FROM (SELECT generate_series(MIN(date_created),
                               MAX(date_created),
                               '1 day')::DATE AS day
          -- What table is date_created in?
          FROM evanston311) AS all_dates
-- 4) Select dates (day from above) that are NOT IN the subquery
 WHERE day NOT IN 
       -- 2) Subquery to select all date_created values as dates
       (SELECT date_created::DATE
          FROM evanston311);

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
6 rows affected.


[(datetime.date(2016, 5, 8),),
 (datetime.date(2016, 11, 6),),
 (datetime.date(2017, 2, 5),),
 (datetime.date(2017, 3, 12),),
 (datetime.date(2018, 1, 6),),
 (datetime.date(2018, 1, 14),)]

### Custom aggregation periods


In [117]:
%%sql

-- Generate 6 month bins covering 2016-01-01 to 2018-06-30

-- Create lower bounds of bins
SELECT generate_series('2016-01-01',  -- First bin lower value
                       '2018-01-01',  -- Last bin lower value
                       '6 months'::interval) AS lower,
-- Create upper bounds of bins
       generate_series('2016-07-01',  -- First bin upper value
                       '2018-07-01',  -- Last bin upper value
                       '6 months'::interval) AS upper;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
5 rows affected.


[(datetime.datetime(2016, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000)))),
 (datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000)))),
 (datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000)))),
 (datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000)))),
 (datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2018, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))))]

In [118]:
%%sql

-- Count number of requests made per day
SELECT day, COUNT(date_created) AS count
-- Use a daily series from 2016-01-01 to 2018-06-30 
-- to include days with no requests
  FROM (SELECT generate_series('2016-01-01',  -- series start date
                               '2018-06-30',  -- series end date
                               '1 day'::interval)::date AS day) AS daily_series
       LEFT JOIN evanston311
       -- match day from above (which is a date) to date_created
       ON day = date_created::DATE
 GROUP BY day;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
912 rows affected.


[(datetime.date(2016, 1, 1), 3),
 (datetime.date(2016, 1, 2), 27),
 (datetime.date(2016, 1, 3), 8),
 (datetime.date(2016, 1, 4), 55),
 (datetime.date(2016, 1, 5), 33),
 (datetime.date(2016, 1, 6), 45),
 (datetime.date(2016, 1, 7), 34),
 (datetime.date(2016, 1, 8), 29),
 (datetime.date(2016, 1, 9), 17),
 (datetime.date(2016, 1, 10), 1),
 (datetime.date(2016, 1, 11), 39),
 (datetime.date(2016, 1, 12), 46),
 (datetime.date(2016, 1, 13), 20),
 (datetime.date(2016, 1, 14), 24),
 (datetime.date(2016, 1, 15), 19),
 (datetime.date(2016, 1, 16), 11),
 (datetime.date(2016, 1, 17), 2),
 (datetime.date(2016, 1, 18), 15),
 (datetime.date(2016, 1, 19), 30),
 (datetime.date(2016, 1, 20), 34),
 (datetime.date(2016, 1, 21), 33),
 (datetime.date(2016, 1, 22), 29),
 (datetime.date(2016, 1, 23), 18),
 (datetime.date(2016, 1, 24), 6),
 (datetime.date(2016, 1, 25), 37),
 (datetime.date(2016, 1, 26), 26),
 (datetime.date(2016, 1, 27), 19),
 (datetime.date(2016, 1, 28), 19),
 (datetime.date(2016, 1, 29), 23),
 (datetime.date(2016, 1, 30), 22),
 (datetime.date(2016, 1, 31), 3),
 (datetime.date(2016, 2, 1), 30),
 (datetime.date(2016, 2, 2), 21),
 (datetime.date(2016, 2, 3), 63),
 (datetime.date(2016, 2, 4), 54),
 (datetime.date(2016, 2, 5), 34),
 (datetime.date(2016, 2, 6), 19),
 (datetime.date(2016, 2, 7), 7),
 (datetime.date(2016, 2, 8), 35),
 (datetime.date(2016, 2, 9), 36),
 (datetime.date(2016, 2, 10), 28),
 (datetime.date(2016, 2, 11), 24),
 (datetime.date(2016, 2, 12), 23),
 (datetime.date(2016, 2, 13), 15),
 (datetime.date(2016, 2, 14), 2),
 (datetime.date(2016, 2, 15), 10),
 (datetime.date(2016, 2, 16), 29),
 (datetime.date(2016, 2, 17), 31),
 (datetime.date(2016, 2, 18), 40),
 (datetime.date(2016, 2, 19), 67),
 (datetime.date(2016, 2, 20), 27),
 (datetime.date(2016, 2, 21), 5),
 (datetime.date(2016, 2, 22), 67),
 (datetime.date(2016, 2, 23), 44),
 (datetime.date(2016, 2, 24), 34),
 (datetime.date(2016, 2, 25), 48),
 (datetime.date(2016, 2, 26), 20),
 (datetime.date(2016, 2, 27), 18),
 (datetime.date(2016, 2, 28), 2),
 (datetime.date(2016, 2, 29), 58),
 (datetime.date(2016, 3, 1), 47),
 (datetime.date(2016, 3, 2), 37),
 (datetime.date(2016, 3, 3), 29),
 (datetime.date(2016, 3, 4), 27),
 (datetime.date(2016, 3, 5), 8),
 (datetime.date(2016, 3, 6), 6),
 (datetime.date(2016, 3, 7), 39),
 (datetime.date(2016, 3, 8), 36),
 (datetime.date(2016, 3, 9), 52),
 (datetime.date(2016, 3, 10), 42),
 (datetime.date(2016, 3, 11), 50),
 (datetime.date(2016, 3, 12), 24),
 (datetime.date(2016, 3, 13), 8),
 (datetime.date(2016, 3, 14), 48),
 (datetime.date(2016, 3, 15), 64),
 (datetime.date(2016, 3, 16), 51),
 (datetime.date(2016, 3, 17), 43),
 (datetime.date(2016, 3, 18), 47),
 (datetime.date(2016, 3, 19), 17),
 (datetime.date(2016, 3, 20), 7),
 (datetime.date(2016, 3, 21), 41),
 (datetime.date(2016, 3, 22), 47),
 (datetime.date(2016, 3, 23), 52),
 (datetime.date(2016, 3, 24), 39),
 (datetime.date(2016, 3, 25), 58),
 (datetime.date(2016, 3, 26), 18),
 (datetime.date(2016, 3, 27), 3),
 (datetime.date(2016, 3, 28), 39),
 (datetime.date(2016, 3, 29), 35),
 (datetime.date(2016, 3, 30), 45),
 (datetime.date(2016, 3, 31), 43),
 (datetime.date(2016, 4, 1), 49),
 (datetime.date(2016, 4, 2), 12),
 (datetime.date(2016, 4, 3), 4),
 (datetime.date(2016, 4, 4), 50),
 (datetime.date(2016, 4, 5), 56),
 (datetime.date(2016, 4, 6), 41),
 (datetime.date(2016, 4, 7), 39),
 (datetime.date(2016, 4, 8), 29),
 (datetime.date(2016, 4, 9), 22),
 (datetime.date(2016, 4, 10), 1),
 (datetime.date(2016, 4, 11), 53),
 (datetime.date(2016, 4, 12), 52),
 (datetime.date(2016, 4, 13), 78),
 (datetime.date(2016, 4, 14), 43),
 (datetime.date(2016, 4, 15), 42),
 (datetime.date(2016, 4, 16), 12),
 (datetime.date(2016, 4, 17), 4),
 (datetime.date(2016, 4, 18), 69),
 (datetime.date(2016, 4, 19), 47),
 (datetime.date(2016, 4, 20), 40),
 (datetime.date(2016, 4, 21), 69),
 (datetime.date(2016, 4, 22), 52),
 (datetime.date(2016, 4, 23), 23),
 (datetime.date(2016, 4, 24), 6),
 (datetime.date(2016, 4, 25), 

In [119]:
%%sql

-- Bins from Step 1
WITH bins AS (
	 SELECT generate_series('2016-01-01',
                            '2018-01-01',
                            '6 months'::interval) AS lower,
            generate_series('2016-07-01',
                            '2018-07-01',
                            '6 months'::interval) AS upper),
-- Daily counts from Step 2
     daily_counts AS (
     SELECT day, count(date_created) AS count
       FROM (SELECT generate_series('2016-01-01',
                                    '2018-06-30',
                                    '1 day'::interval)::date AS day) AS daily_series
            LEFT JOIN evanston311
            ON day = date_created::date
      GROUP BY day)
-- Select bin bounds 
SELECT lower, 
       upper, 
       -- Compute median of count for each bin
       percentile_disc(0.5) WITHIN GROUP (ORDER BY count) AS median
  -- Join bins and daily_counts
  FROM bins
       LEFT JOIN daily_counts
       -- Where the day is between the bin bounds
       ON day >= lower
          AND day <= upper
 -- Group by bin bounds
 GROUP BY lower, upper
 ORDER BY lower;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
5 rows affected.


[(datetime.datetime(2016, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 37),
 (datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 41),
 (datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 44),
 (datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 51),
 (datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2018, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 40)]

### Monthly average with missing dates


In [120]:
%%sql

-- generate series with all days from 2016-01-01 to 2018-06-30
WITH all_days AS 
     (SELECT generate_series('2016-01-01',
                             '2018-06-30',
                             '1 day'::interval) AS date),
     -- Subquery to compute daily counts
     daily_count AS 
     (SELECT date_trunc('day', date_created) AS day,
             count(*) AS count
        FROM evanston311
       GROUP BY day)
-- Aggregate daily counts by month using date_trunc
SELECT date_trunc('month', date) AS month,
       -- Use coalesce to replace NULL count values with 0
       avg(coalesce(count, 0)) AS average
  FROM all_days
       LEFT JOIN daily_count
       -- Joining condition
       ON all_days.date=daily_count.day
 GROUP BY month
 ORDER BY month; 

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
30 rows affected.


[(datetime.datetime(2016, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('23.4516129032258065')),
 (datetime.datetime(2016, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('30.7241379310344828')),
 (datetime.datetime(2016, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.5483870967741935')),
 (datetime.datetime(2016, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('37.3000000000000000')),
 (datetime.datetime(2016, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('39.4516129032258065')),
 (datetime.datetime(2016, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('44.0000000000000000')),
 (datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('41.4838709677419355')),
 (datetime.datetime(2016, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('46.6129032258064516')),
 (datetime.datetime(2016, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('47.2666666666666667')),
 (datetime.datetime(2016, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.8064516129032258')),
 (datetime.datetime(2016, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.4333333333333333')),
 (datetime.datetime(2016, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('29.3548387096774194')),
 (datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('34.9032258064516129')),
 (datetime.datetime(2017, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('31.5357142857142857')),
 (datetime.datetime(2017, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('34.4838709677419355')),
 (datetime.datetime(2017, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('42.2000000000000000')),
 (datetime.datetime(2017, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('46.8064516129032258')),
 (datetime.datetime(2017, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('69.4666666666666667')),
 (datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('57.3225806451612903')),
 (datetime.datetime(2017, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('53.7419354838709677')),
 (datetime.datetime(2017, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('44.6666666666666667')),
 (datetime.datetime(2017, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('41.5806451612903226')),
 (datetime.datetime(2017, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('40.7000000000000000')),
 (datetime.datetime(2017, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.0967741935483871')),
 (datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('33.1935483870967742')),
 (datetime.datetime(2018, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('30.5714285714285714')),
 (datetime.datetime(2018, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('29.6774193548387097')),
 (datetime.datetime(2018, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('35.1333333333333333')),
 (datetime.datetim

## Time between events
### Longest gap


In [122]:
%%sql

-- Compute the gaps
WITH request_gaps AS (
        SELECT date_created,
               -- lead or lag
               lag(date_created) OVER (ORDER BY date_created) AS previous,
               -- compute gap as date_created minus lead or lag
               date_created - lag(date_created) OVER (ORDER BY date_created) AS gap
          FROM evanston311)
-- Select the row with the maximum gap
SELECT *
  FROM request_gaps
-- Subquery to select maximum gap from request_gaps
 WHERE gap = (SELECT MAX(gap)
                FROM request_gaps);

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(datetime.datetime(2018, 1, 7, 14, 41, 34, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.datetime(2018, 1, 5, 14, 4, 9, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), datetime.timedelta(days=2, seconds=2245))]

### Rats

In [124]:
%%sql

-- Truncate the time to complete requests to the day
SELECT date_trunc('day', date_completed - date_created) AS completion_time,
-- Count requests with each truncated time
       COUNT(date_completed)
  FROM evanston311
-- Where category is rats
 WHERE category = 'Rodents- Rats'
-- Group and order by the variable of interest
 GROUP BY completion_time
 ORDER BY count
LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[(datetime.timedelta(days=239), 1),
 (datetime.timedelta(days=153), 1),
 (datetime.timedelta(days=291), 1),
 (datetime.timedelta(days=263), 1),
 (datetime.timedelta(days=157), 1),
 (datetime.timedelta(days=189), 1),
 (datetime.timedelta(days=314), 1),
 (datetime.timedelta(days=204), 1),
 (datetime.timedelta(days=238), 1),
 (datetime.timedelta(days=190), 1),
 (datetime.timedelta(days=348), 1),
 (datetime.timedelta(days=286), 1),
 (datetime.timedelta(days=214), 1),
 (datetime.timedelta(days=177), 1),
 (datetime.timedelta(days=283), 1),
 (datetime.timedelta(days=280), 1),
 (datetime.timedelta(days=148), 1),
 (datetime.timedelta(days=170), 1),
 (datetime.timedelta(days=161), 1),
 (datetime.timedelta(days=219), 1)]

In [126]:
%%sql

SELECT category, 
       -- Compute average completion time per category
       avg(date_completed - date_created) AS avg_completion_time
  FROM evanston311
-- Where completion time is less than the 95th percentile value
 WHERE date_completed - date_created < 
-- Compute the 95th percentile of completion time in a subquery
         (SELECT percentile_disc(0.95) WITHIN GROUP (ORDER BY date_completed - date_created)
            FROM evanston311)
 GROUP BY category
-- Order the results
 ORDER BY avg_completion_time DESC
 LIMIT 20;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
20 rows affected.


[('Trash Cart - Downsize, Upsize or Remove', datetime.timedelta(days=12, seconds=64070, microseconds=586912)),
 ('Sanitation Billing Questions', datetime.timedelta(days=12, seconds=40405, microseconds=888889)),
 ('THIS REQUEST IS INACTIVE...Trash Cart - Compost Bin', datetime.timedelta(days=12, seconds=23562, microseconds=24390)),
 ('Trash, Recycling, Yard Waste Cart- Repair/Replacement', datetime.timedelta(days=11, seconds=67707, microseconds=488108)),
 ('Rodents- Rats', datetime.timedelta(days=11, seconds=32280, microseconds=840849)),
 ('Landmark Building Plaque or Nomination', datetime.timedelta(days=11, seconds=18671, microseconds=666667)),
 ('Notice of Violation', datetime.timedelta(days=10, seconds=70129, microseconds=142858)),
 ('Move Out- Special Pickup', datetime.timedelta(days=10, seconds=28620, microseconds=500000)),
 ('Long Term Care', datetime.timedelta(days=9, seconds=83529)),
 ('Smoking', datetime.timedelta(days=8, seconds=83706, microseconds=611111)),
 ('Trash - Special Pickup (STAFF ONLY)', datetime.timedelta(days=8, seconds=43442, microseconds=98361)),
 ('Trash - Special Pickup - Resident Use', datetime.timedelta(days=7, seconds=53594, microseconds=687500)),
 ('Bike Racks', datetime.timedelta(days=7, seconds=36291, microseconds=230769)),
 ('Accessibility', datetime.timedelta(days=7, seconds=25009)),
 ('Abandoned Bicycle on City Property', datetime.timedelta(days=7, seconds=17212, microseconds=560606)),
 ('Refuse Cart Placement', datetime.timedelta(days=7, seconds=14458)),
 ('Post Office- Question or Concern', datetime.timedelta(days=6, seconds=61777, microseconds=333333)),
 ('General/Routine Maintenance - Facilities Management', datetime.timedelta(days=6, seconds=61590, microseconds=700000)),
 ('General/Routine Maintenance - Facilities (FIRE STATIONS ONLY)', datetime.timedelta(days=6, seconds=47541)),
 ('Public Transit Agency Issue', datetime.timedelta(days=6, seconds=44007, microseconds=150000))]

In [127]:
%%sql

-- Compute correlation (corr) between 
-- avg_completion time and count from the subquery
SELECT corr(avg_completion, count)
  -- Convert date_created to its month with date_trunc
  FROM (SELECT date_trunc('month', date_created) AS month, 
               -- Compute average completion time in number of seconds           
               AVG(EXTRACT(epoch FROM date_completed - date_created)) AS avg_completion, 
               -- Count requests per month
               count(*) AS count
          FROM evanston311
         -- Limit to rodents
         WHERE category='Rodents- Rats' 
         -- Group by month, created above
         GROUP BY month) 
         -- Required alias for subquery 
         AS monthly_avgs;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
1 rows affected.


[(0.23358132177837962,)]

In [128]:
%%sql

-- Compute monthly counts of requests created
WITH created AS (
       SELECT date_trunc('month', date_created) AS month,
              count(*) AS created_count
         FROM evanston311
        WHERE category='Rodents- Rats'
        GROUP BY month),
-- Compute monthly counts of requests completed
      completed AS (
       SELECT date_trunc('month', date_completed) AS month,
              count(*) AS completed_count
         FROM evanston311
        WHERE category='Rodents- Rats'
        GROUP BY month)
-- Join monthly created and completed counts
SELECT created.month, 
       created_count, 
       completed_count
  FROM created
       INNER JOIN completed
       ON created.month=completed.month
 ORDER BY created.month;

 * postgresql+psycopg2://adrik:***@localhost:5432/adrik
30 rows affected.


[(datetime.datetime(2016, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 11, 1),
 (datetime.datetime(2016, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 21, 11),
 (datetime.datetime(2016, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 31, 14),
 (datetime.datetime(2016, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 36, 16),
 (datetime.datetime(2016, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 40, 19),
 (datetime.datetime(2016, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 41, 49),
 (datetime.datetime(2016, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 80, 47),
 (datetime.datetime(2016, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 80, 43),
 (datetime.datetime(2016, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 55, 58),
 (datetime.datetime(2016, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 76, 67),
 (datetime.datetime(2016, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 63, 61),
 (datetime.datetime(2016, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 20, 83),
 (datetime.datetime(2017, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 16, 34),
 (datetime.datetime(2017, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 20, 30),
 (datetime.datetime(2017, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 24, 32),
 (datetime.datetime(2017, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 43, 14),
 (datetime.datetime(2017, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 50, 17),
 (datetime.datetime(2017, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 65, 34),
 (datetime.datetime(2017, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 68, 46),
 (datetime.datetime(2017, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 87, 53),
 (datetime.datetime(2017, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 54, 70),
 (datetime.datetime(2017, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 78, 12),
 (datetime.datetime(2017, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 44, 174),
 (datetime.datetime(2017, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 12, 65),
 (datetime.datetime(2018, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 19, 45),
 (datetime.datetime(2018, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 10, 10),
 (datetime.datetime(2018, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 28, 27),
 (datetime.datetime(2018, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 35, 23),
 (datetime.datetime(2018, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 46, 47),
 (datetime.datetime(2018, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), 52, 55)]